pkg/bucket/bandwidth/monitor.go

```diff
@@ -0,0 +1,186 @@
+/*
+ * MinIO Cloud Storage, (C) 2020 MinIO, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package bandwidth
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/minio/minio/pkg/bucket/bandwidth/measurements"
+	"github.com/minio/minio/pkg/bucket/bandwidth/token"
+	"github.com/minio/minio/pkg/pubsub"
+)
+
+const maxGoRoutines = 4
+
+// Update captures the details for the data transferred per object.
+type Update struct {
+	Done        bool      // Is this the last update for an object
+	BucketToken int       // Token for bucket, used to avoid string copies per update
+	ObjectToken int       // Token for object, used to avoid strong copies per update
+	Bytes       uint64    // Size of the latest update
+	Start       time.Time // When did the transfer start
+	Stop        time.Time // When did the transfer stop
+}
+
+// Track returns the tokens used to track the bucket, object tuple.
+func (m *Monitor) Track(bucket string, object string) (int, int) {
+	m.lock.Lock()
+	m.lock.Unlock()
+	o := m.tokenizer.GetToken(bucket)
+	b := m.tokenizer.GetToken(objectKeyForBucket(bucket, object))
+	return o, b
+}
+
+// Details for the measured bandwidth
+type Details struct {
+	AggregateBandwidth float64 `json:"aggregateBandwidth"`
+	MovingBandwidth    float64 `json:"currentBandwidth"`
+}
+
+// Report captures the details for all buckets.
+type Report struct {
+	BucketStats map[string]Details `json:"bucketStats,omitempty"`
+}
+
+// SubscribeAllBuckets subscribes to all the buckets monitored
+func (m *Monitor) SubscribeAllBuckets(subCh chan interface{}, doneCh <-chan struct{}) {
```

Comment:
Use context for cancelation.

Reply:
Will do

Reply:
Yes. The caller sends a context. When it no longer is interested it will cancel it.